### PR TITLE
Add destructible component threshold triggers

### PR DESCRIPTION
**Pull request by DrSmugleaf**
Needed for https://github.com/space-wizards/space-station-14/pull/2783
Needed for https://github.com/space-wizards/space-station-14/pull/2022

## About the PR 
Previously thresholds were triggered when going above a certain amount of total damage.
This PR moves that logic into classes that implement ITrigger which are then specified through YAML, adding support for thresholds that trigger on specific amounts of classes of damage (brute/burn) and types of damage (blunt/slash).

For example:
```
  components:
  - type: Damageable
  - type: Destructible
    thresholds:
    - trigger:
        !type:TotalDamageClassesTrigger
        damage:
          Brute: 10
          Burn: 10
```

This will add a threshold that triggers when the entity reaches 10 brute and 10 burn damage.
If the damage keeps increasing past that, it will not trigger again.
If any of the two damage types is healed below 10 and then increased again, it will trigger once again only if TriggersOnce is set to false (currently the default) in the Threshold class.
A threshold with TriggersOnce set to false will not trigger again until its trigger has returned false at least one time since it was last activated, to prevent repeat activations.

The same can be done with damage types by using TotalDamageTypesTrigger.

Previously, thresholds were defined in YAML like this:
```
  - type: Destructible
    thresholds:
      500:
        behaviors:
        - !type:DoActsBehavior
          acts: ["Destruction"]
```

The same YAML would now be specified like this, for a trigger that activates at 500 total damage:
```
  - type: Destructible
    thresholds:
    - trigger:
        !type:TotalDamageTrigger
        damage: 500
      behaviors:
      - !type:DoActsBehavior
        acts: ["Destruction"]
```

Each threshold is checked in the order that it was defined in YAML.
Each threshold may only have one trigger (with the possibility of adding composite or more complicated triggers in the future). 
This would allow for random triggers, triggers dependent on the world, etc.
Each threshold may have any amount of behaviors that are executed in the order defined in YAML as done by Vera in https://github.com/space-wizards/space-station-14/pull/2892.
Triggers do not store their previous state themselves, that is done by Threshold with OldTriggered,

This does not update saltern or any other maps.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Destructible/Thresholds/Behaviors/IBehavior.cs [hunk 1]
```diff
@@ -2,9 +2,9 @@
 using Robust.Shared.Interfaces.GameObjects;
 using Robust.Shared.Interfaces.Serialization;
 
-namespace Content.Server.GameObjects.Components.Destructible.Thresholds.Behavior
+namespace Content.Server.GameObjects.Components.Destructible.Thresholds.Behaviors
 {
-    public interface IThresholdBehavior : IExposeData
+    public interface IBehavior : IExposeData
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Destructible/Thresholds/Behaviors/IBehavior.cs:7]: Could you please rename this again to something else? `IBehavior` alone might be real confusing given we also have `IActionBehavior, IMechanismBehavior, IItemActionBehavior`...

File: Content.Server/GameObjects/Components/Destructible/Thresholds/Triggers/ITrigger.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using Content.Server.GameObjects.EntitySystems;
+using Content.Shared.GameObjects.Components.Damage;
+using Robust.Shared.Interfaces.Serialization;
+
+namespace Content.Server.GameObjects.Components.Destructible.Thresholds.Triggers
+{
+    public interface ITrigger : IExposeData
```
Comment (Zumorica) [inline Content.Server/GameObjects/Components/Destructible/Thresholds/Triggers/ITrigger.cs:7]: Same here, honestly. Would be best if it had a name that makes it clear this relates to destructible thresholds and stuff.

File: Resources/Prototypes/Entities/Mobs/Species/human.yml [hunk 3]
```diff
@@ -164,9 +164,11 @@
       200: !type:DeadMobState {}
   - type: Destructible
     thresholds:
-      400:
-        behaviors:
-          - !type:GibBehavior { }
+    - trigger:
+        !type:TotalDamageTrigger
```
Comment (Zumorica) [inline Resources/Prototypes/Entities/Mobs/Species/human.yml:168]: While you're at it, could you make this only happen with blunt damage? (although even THAT is still placeholder, as gib shouldn't be triggered by a certain amount of damage, but...)

File: Content.Server/GameObjects/Components/Destructible/Thresholds/Behaviors/IBehavior.cs (hunk 1 above)
Comment (DrSmugleaf) [reply Content.Server/GameObjects/Components/Destructible/Thresholds/Behaviors/IBehavior.cs]: The naming is something I can't quite decide on, between Thresholds, Triggers and Behaviors. I removed the Threshold prefix as they could be used for something else but given that they accept DestructibleSystem as an argument for dependencies I agree with naming it IThresholdBehavior again. Fixed.

File: Content.Server/GameObjects/Components/Destructible/Thresholds/Triggers/ITrigger.cs (hunk 2 above)
Comment (DrSmugleaf) [reply Content.Server/GameObjects/Components/Destructible/Thresholds/Triggers/ITrigger.cs]: Fixed in https://github.com/space-wizards/space-station-14/pull/3080/commits/c99be52304001825c5c0259f156f46fbca067c2d. See https://github.com/space-wizards/space-station-14/pull/3080#discussion_r570507140

File: Resources/Prototypes/Entities/Mobs/Species/human.yml (hunk 3 above)
Comment (DrSmugleaf) [reply Resources/Prototypes/Entities/Mobs/Species/human.yml]: Done in https://github.com/space-wizards/space-station-14/pull/3080/commits/d51ce1dd25022d29763525b12fa51c976b249072

**Review by Zumorica — APPROVED** [review]
Looks good.

**Review by Zumorica — APPROVED** [review]
Just tested ingame as well, nothing seems broken and everything works fine :+1:

**Pull request merged**